File: android/app/src/jvmShared/kotlin/com/openclaw/app/ui/ChatModel.kt
package com.openclaw.app.ui

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshots.SnapshotStateList
import org.json.JSONArray
import org.json.JSONObject

/** Marker the agent emits to hand a file/photo back to the user (see backend `_FILE_PROTOCOL`). */
val FILE_MARKER = Regex("""\[\[FILE:([^\]]+)]]""")

/** A piece of an assistant message: streamed text, or a tool (terminal) block. */
sealed interface Seg

class TextSeg(initial: String = "") : Seg {
    var text by mutableStateOf(initial)
}

class ToolSeg(val command: String) : Seg {
    var output by mutableStateOf("")
    var exit by mutableStateOf<Int?>(null)
}

/** An attachment: a file sent by the user (localPath set) or returned by the agent
 *  (remotePath set, fetched on demand). Images render inline. */
class FileSeg(val name: String, val remotePath: String, val isImage: Boolean) : Seg {
    var localPath by mutableStateOf<String?>(null) // present once available locally (sent, or downloaded)
    var downloading by mutableStateOf(false)
    var error by mutableStateOf<String?>(null)
}

/** An interactive multiple-choice question the agent asked (AskUserQuestion). */
data class QOption(val label: String, val description: String)
data class Q(val question: String, val header: String, val multiSelect: Boolean, val options: List<QOption>)
class QuestionSeg(val id: String, val questions: List<Q>) : Seg {
    var answered by mutableStateOf(false)
}

/** Parse the AskUserQuestion `questions` payload into UI models. */
fun parseQuestions(arr: JSONArray): List<Q> {
    val out = ArrayList<Q>()
    for (i in 0 until arr.length()) {
        val o = arr.optJSONObject(i) ?: continue
        val opts = ArrayList<QOption>()
        val oarr = o.optJSONArray("options") ?: JSONArray()
        for (j in 0 until oarr.length()) {
            val oo = oarr.optJSONObject(j) ?: continue
            opts.add(QOption(oo.optString("label"), oo.optString("description")))
        }
        out.add(Q(o.optString("question"), o.optString("header"), o.optBoolean("multiSelect"), opts))
    }
    return out
}

class ChatMsg(val role: String) {
    val segs: SnapshotStateList<Seg> = mutableStateListOf()
    var streaming by mutableStateOf(false)
    var error by mutableStateOf<String?>(null)

    fun appendText(s: String) {
        val last = segs.lastOrNull()
        if (last is TextSeg) last.text += s
        else segs.add(TextSeg(s))
    }

    fun addTool(command: String) {
        segs.add(ToolSeg(command))
    }

    fun fillTool(exitCode: Int, out: String) {
        (segs.lastOrNull { it is ToolSeg } as? ToolSeg)?.let {
            it.exit = exitCode
            it.output = out
        }
    }

    /** Plain-text view (for the user bubble / simple display). */
    fun plain(): String = segs.filterIsInstance<TextSeg>().joinToString("") { it.text }
}

fun userMsg(text: String): ChatMsg = ChatMsg("user").apply { segs.add(TextSeg(text)) }
fun assistantMsg(): ChatMsg = ChatMsg("assistant").apply { streaming = true }

/** Serialize a finished message for server-side history (tool output capped to keep it small). */
fun ChatMsg.toJson(): JSONObject {
    val arr = JSONArray()
    segs.forEach { seg ->
        val o = JSONObject()
        when (seg) {
            is TextSeg -> { o.put("t", "text"); o.put("text", seg.text) }
            is ToolSeg -> {
                o.put("t", "tool"); o.put("command", seg.command)
                o.put("output", seg.output.take(4000)); seg.exit?.let { o.put("exit", it) }
            }
            is FileSeg -> { o.put("t", "file"); o.put("name", seg.name); o.put("path", seg.remotePath); o.put("image", seg.isImage) }
            // History keeps questions as plain text; the answer is its own user message.
            is QuestionSeg -> { o.put("t", "text"); o.put("text", seg.questions.joinToString("\n") { "❓ ${it.question}" }) }
        }
        arr.put(o)
    }
    val m = JSONObject().put("role", role).put("segs", arr)
    error?.let { m.put("error", it) }
    return m
}

/** Rebuild a message from stored history (non-streaming; images re-fetched on demand). */
fun chatMsgFromJson(o: JSONObject): ChatMsg {
    val m = ChatMsg(o.optString("role", "assistant"))
    val arr = o.optJSONArray("segs") ?: JSONArray()
    for (i in 0 until arr.length()) {
        val s = arr.optJSONObject(i) ?: continue
        when (s.optString("t")) {
            "text" -> m.segs.add(TextSeg(s.optString("text")))
            "tool" -> m.segs.add(ToolSeg(s.optString("command")).apply {
                output = s.optString("output")
                if (s.has("exit") && !s.isNull("exit")) exit = s.getInt("exit")
            })
            "file" -> m.segs.add(FileSeg(s.optString("name"), s.optString("path"), s.optBoolean("image")))
        }
    }
    m.streaming = false
    o.optString("error").takeIf { it.isNotBlank() }?.let { m.error = it }
    return m
}

File: ios/ClawNest/Model/ChatModel.swift
import Foundation

/// Marker the agent emits to hand a file/photo back to the user (see backend `_FILE_PROTOCOL`).
let FILE_MARKER = try! NSRegularExpression(pattern: #"\[\[FILE:([^\]]+)\]\]"#)

private let IMAGE_EXT = try! NSRegularExpression(
    pattern: #"(?i).*\.(png|jpe?g|gif|webp|bmp|heic|heif)$"#)

func isImageName(_ name: String) -> Bool {
    let r = NSRange(name.startIndex..., in: name)
    return IMAGE_EXT.firstMatch(in: name, range: r) != nil
}

/// A piece of an assistant/user message. Flat, mutable value type (mirrors ChatModel.kt's Seg
/// hierarchy). `kind` selects which fields are meaningful.
struct Seg: Identifiable, Hashable {
    enum Kind: String, Hashable { case text, tool, file, question }
    let id = UUID()
    var kind: Kind

    // text
    var text: String = ""
    // tool
    var command: String = ""
    var output: String = ""
    var exit: Int? = nil
    // file
    var fileName: String = ""
    var remotePath: String = ""
    var isImage: Bool = false
    var localPath: String? = nil
    var downloading: Bool = false
    var fileError: String? = nil
    // question
    var questionId: String = ""
    var questions: [Q] = []
    var answered: Bool = false

    static func text(_ s: String) -> Seg { Seg(kind: .text, text: s) }
    static func tool(_ cmd: String) -> Seg { Seg(kind: .tool, command: cmd) }
    static func file(name: String, remotePath: String, isImage: Bool, localPath: String? = nil) -> Seg {
        var s = Seg(kind: .file); s.fileName = name; s.remotePath = remotePath
        s.isImage = isImage; s.localPath = localPath; return s
    }
    static func question(id: String, questions: [Q]) -> Seg {
        var s = Seg(kind: .question); s.questionId = id; s.questions = questions; return s
    }
}

/// One chat message. ObservableObject so streaming mutations re-render just this bubble.
final class ChatMsg: ObservableObject, Identifiable {
    let id = UUID()
    let role: String            // "user" | "assistant"
    @Published var segs: [Seg]
    @Published var streaming: Bool
    @Published var error: String?

    init(role: String, segs: [Seg] = [], streaming: Bool = false, error: String? = nil) {
        self.role = role; self.segs = segs; self.streaming = streaming; self.error = error
    }

    func appendText(_ s: String) {
        if var last = segs.last, last.kind == .text {
            last.text += s
            segs[segs.count - 1] = last
        } else {
            segs.append(.text(s))
        }
    }

    func addTool(_ command: String) { segs.append(.tool(command)) }

    func fillTool(exit: Int, output: String) {
        if let idx = segs.lastIndex(where: { $0.kind == .tool }) {
            segs[idx].exit = exit
            segs[idx].output = output
        }
    }

    /// Plain-text view (user bubble / dedupe of proactive pushes).
    func plain() -> String { segs.filter { $0.kind == .text }.map(\.text).joined() }

    // ---- history serialization (mirrors ChatMsg.toJson / chatMsgFromJson) ----

    func toJson() -> [String: Any] {
        var arr: [[String: Any]] = []
        for seg in segs {
            switch seg.kind {
            case .text:
                arr.append(["t": "text", "text": seg.text])
            case .tool:
                var o: [String: Any] = ["t": "tool", "command": seg.command,
                                        "output": String(seg.output.prefix(4000))]
                if let e = seg.exit { o["exit"] = e }
                arr.append(o)
            case .file:
                arr.append(["t": "file", "name": seg.fileName, "path": seg.remotePath, "image": seg.isImage])
            case .question:
                let joined = seg.questions.map { "❓ \($0.question)" }.joined(separator: "\n")
                arr.append(["t": "text", "text": joined])
            }
        }
        var m: [String: Any] = ["role": role, "segs": arr]
        if let error { m["error"] = error }
        return m
    }

    static func fromJson(_ o: [String: Any]) -> ChatMsg {
        let m = ChatMsg(role: o["role"] as? String ?? "assistant")
        let arr = o["segs"] as? [[String: Any]] ?? []
        for s in arr {
            switch s["t"] as? String {
            case "text": m.segs.append(.text(s["text"] as? String ?? ""))
            case "tool":
                var seg = Seg.tool(s["command"] as? String ?? "")
                seg.output = s["output"] as? String ?? ""
                if let e = s["exit"] as? Int { seg.exit = e }
                m.segs.append(seg)
            case "file":
                m.segs.append(.file(name: s["name"] as? String ?? "",
                                    remotePath: s["path"] as? String ?? "",
                                    isImage: s["image"] as? Bool ?? false))
            default: break
            }
        }
        m.streaming = false
        if let e = o["error"] as? String, !e.isEmpty { m.error = e }
        return m
    }
}

func userMsg(_ text: String) -> ChatMsg { ChatMsg(role: "user", segs: [.text(text)]) }
func assistantMsg() -> ChatMsg { ChatMsg(role: "assistant", streaming: true) }
